File: src/src_core/fmg/common/geom/DoubleExt.cs
using System;

namespace fmg.common.geom {

    public static class DoubleExt {

        /** Equals N digit precision */
        public static bool EqualsPrecision(this double value1, double value2, double precision = 0.00001) {
            return Math.Abs(value1 - value1) <= precision;
        }

        /// <summary> Минимально различны? Has minimal differences
        /// https://msdn.microsoft.com/en-us/library/ya2zha7s%28v=vs.110%29.aspx </summary>
        public static bool HasMinDiff(this double value1, double value2, uint units = 1) {
            //return Math.Abs(value1 - value2) <= (1.0/ (10 * units));
            var lValue1 = BitConverter.DoubleToInt64Bits(value1);
            var lValue2 = BitConverter.DoubleToInt64Bits(value2);

            // If the signs are different, return false except for +0 and -0.
            if ((lValue1 >> 63) != (lValue2 >> 63)) {
                return value1.Equals(value2);
            }

            var diff = Math.Abs(lValue1 - lValue2);
            return (diff <= units);
        }

    }

}

File: FastMines_core/src/fmg/common/geom/DoubleExt.java
package fmg.common.geom;

public final class DoubleExt {
    private DoubleExt() {}

    /** Equals N digit precision */
    public static boolean equalsPrecision(double value1, double value2) {
        return equalsPrecision(value1, value2, 0.00001);
    }

    public static boolean equalsPrecision(double value1, double value2, double precision) {
        return Math.abs(value1 - value2) <= precision;
    }

    /** @return value1 == value2 */
    public static boolean almostEquals(double value1, double value2) {
        return almostEquals(value1, value2, 1);
    }

    public static boolean almostEquals(double value1, double value2, int units) {
        long lValue1 = Double.doubleToLongBits(value1);
        long lValue2 = Double.doubleToLongBits(value2);

        // If the signs are different, return false except for +0 and -0.
        if ((lValue1 >>> 63) != (lValue2 >>> 63)) {
            return 0 == Double.compare(lValue1, value2);
        }

        double diff = Math.abs(lValue1 - lValue2);
        return (diff <= units);
    }

}
